---
title: ChIP-Seq 1 Project Report 
author: "Arafat Rahman (arahm010@ucr.edu)"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`" 
output:
    html_document:
    toc: true
    toc_float:
        collapsed: true
        smooth_scroll: true
    toc_depth: 3
    fig_caption: yes
    code_folding: show
    number_sections: true

fontsize: 14pt
bibliography: bibtex.bib
---

<!--
- Render from R
rmarkdown::render('ChIPSeq2_arafat_report.Rmd',output_file='report_arafat.html', c('html_document'), clean=F, output_options =list())
-->

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
options(width=100, max.print=1000)
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")))
```


# Abstract

ChIP-Seq is NGS sequencing of transcription factor binding sites in DNA. This experiment can reveal binding sites of particular protein in the genome, and find potential motifs. ChIP-Seq can help to find how a protein interact with particular genetic pathway by regulating transcription-machinery in the cell. In this work, ChIP-Seq dataset from previously published paper were used for educational-purpose data-analysis. The analysis corroborate with the paper. 


# 1. Introduction

Arabidopsis is a flowering plant that can be bred sexually. This is a good model organism for plant development. It breed quickly and are easy to work with. It's pollination is important, because 75% of worldâ€™s leading food crop rely on animal pollination for high yield. AP1 has been shown to suppress other genes that suppress the production of flowers. AP1 regulate suppressors of floral production by MADS-domain transcription factors. Research on Kaufmann et al. 2010 [@kaufmann2010orchestration] paper went into exploring where AP1 binds and whether it upregulates or downregulates the genes that it binds.


Many genes had different transcription levels when AP1 lacks protein biosynthesis-- demonstrating the protein is sufficient to regulate floral development. ap1 mutants demonstrated that the protein is necessary for regulating some high-confidence target genes. It influences transcription factor-coding genes, showing that it regulates transcriptional regulators to ultimately influence floral phenotype. They found that, AP1 preferentially bound prior to the start site of genes that it regulates. AP1 upregulates some genes and downregulates other (mostly it downregulates genes). There are time-differentials in terms of the regulation of genes. AP1 sets up floral meristem identity by suppressing genes that cause the development of shoots.

We have used publication by Kaufmann et al. 2010 paper for our experimental data source [@kaufmann2010orchestration].


# 2. Methods

## 2.1 Experimental design

In the Kaufmann et al. 2010 paper, they have two condigions for ChIP-Seq experiment: pull-down by AP1 specific antibody and control experiment. They have both biological and technical replicates, thereby finally they have seven libraries: three for AP1 treatment and four for control experiments. They parallely also did microarray analysis to record transcriptional change during this modification. 

## 2.2 ChIP Seq Analysis

For our analysis, we have done the following:

* Establish workflow environment
* FastQ quality report
* Use Bowtie2 to map reads to the genome
* Generate BAM files and view them using Integrative Genome Viewer
* Model-based Analsysis for ChIP-Seq 2 (MACS2) for Peak Calling-evaluates the significance of binding regions in ChIP-Seq. 
* Annotating with ChIPseeker package
* Generate count tables
* And challenge project tasks

# 3. Results and Discussion

## 3.1 Generate workflow environment

The workfolw generated by using the starter code from the GitHub project repository. I have used the `git pull` command for initiating local instance in HPCC Biocluster. The workflow requires to load `systemPipeR` package first.

```{r load_systempiper, eval=TRUE}
library(systemPipeR)
```


## 3.2 Read preprocessing

For initial analysis, I have filtered data so that there is no read sequence in the fastq file that contain any base with Phred quality less than 20. But, this filtering changes data-set size roughly by 80%, which is great loss of information. In second round of the analysis, I have used unfiltered data with no read preprocessing.

### 3.2.1 Experiment definition provided by `targets` file

The `targets` file defines all FASTQ files and sample comparisons of the analysis workflow. The target file is populated with 'untrimmed' data-set. The target file is shown below.

```{r load_targets_file, eval=TRUE}
targets <- read.delim("targets_chip.txt", comment.char = "#")
targets
```

### 3.2.2 FASTQ quality report

The following `seeFastq` and `seeFastqPlot` functions generate and plot a series of useful quality
statistics for a set of FASTQ files including per cycle quality box
plots, base proportions, base-level quality trends, relative k-mer
diversity, length and occurrence distribution of reads, number of reads
above quality cutoffs and mean quality distribution. The results are
written to a PDF file named `fastqReport.pdf`.

The report shows overall low quality of the fastq files. As this is one of the first experiment when the NGS sequencing was not matured enough, which explains it's low quality. 

```{r fastq_report, eval=FALSE}
args <- systemArgs(sysma="param/tophat.param", mytargets="targets_chip.txt")
library(BiocParallel); library(BatchJobs)
f <- function(x) {
    library(systemPipeR)
    args <- systemArgs(sysma="param/tophat.param", mytargets="targets_chip.txt")
    seeFastq(fastq=infile1(args)[x], batchsize=100000, klength=8)
}
funs <- makeClusterFunctionsSLURM("slurm.tmpl")
param <- BatchJobsParam(length(args), resources=list(walltime="00:20:00", ntasks=1, ncpus=1, memory="2G"), cluster.functions=funs)
register(param)
fqlist <- bplapply(seq(along=args), f)
#pdf("./results/fastqReport.pdf", height=18, width=4*length(fqlist))
#seeFastqPlot(unlist(fqlist, recursive=FALSE)) 
#dev.off()
png("fastqReport.png", height=18, width=4*length(fqlist), units="in", res=300)
seeFastqPlot(unlist(fqlist, recursive=FALSE))
dev.off()
``` 

![](fastqReport.png)
<div>FASTQ quality report for 7 samples</div></br>


## 3.3 Alignments

### 3.3.1 Read mapping with `Bowtie2` 

The original paper used SOAP program. Here, I have run the alignments using `Bowtie2` sequentially on a single system. TAIR10 was used as reference genome to map the reads.

```{r bowtie2_align_seq, eval=FALSE}
args <- systemArgs(sysma="param/bowtieSE.param", mytargets="targets_chip.txt")
sysargs(args)[1] # Command-line parameters for first FASTQ file
moduleload(modules(args)) # Skip if a module system is not used
system("bowtie2-build ./data/tair10.fasta ./data/tair10.fasta") # Indexes reference genome
runCommandline(args)
```

Check whether all BAM files have been created
```{r check_files_exist, eval=FALSE}
file.exists(outpaths(args))
```

### 3.3.2 Read and alignment stats

The following provides an overview of the number of reads in each sample
and how many of them aligned to the reference. The report shows around 70% reads in average were successfully aligned on the reference genome. 

```{r align_stats, eval=FALSE}
read_statsDF <- alignStats(args=args) 
write.table(read_statsDF, "results/alignStats.xls", row.names=FALSE, quote=FALSE, sep="\t")
```
```{r aligne_stats_show, eval=TRUE}
read.delim("results/alignStats.xls")
```

### 3.3.3 Create symbolic links for viewing BAM files in IGV

Symbolic link to the bam files created so that these can be publically accessible using Internet. The files can be found using this [link:](http://biocluster.ucr.edu/~arahm010/ChIP1_arf/)

```{r symbol_links, eval=FALSE}
symLink2bam(sysargs=args, htmldir=c("~/.html/", "ChIP1_arf/"), 
            urlbase="http://biocluster.ucr.edu/~arahm010/", 
            urlfile="./results/IGVurl.txt```
```


## 3.4 Utilities for coverage data

The following introduces several utilities useful for ChIP-Seq data. They are not part of the actual
workflow.

### 3.4.1 Rle object stores coverage information

Here we created run-length encodoing object to compress coverage information in lossless way.

```{r rle_object, eval=FALSE}
library(rtracklayer); library(GenomicRanges); library(Rsamtools); library(GenomicAlignments)
aligns <- readGAlignments(outpaths(args)[1])
cov <- coverage(aligns)
cov
```

### 3.4.2 Resizing aligned reads
```{r resize_align, eval=FALSE}
trim(resize(as(aligns, "GRanges"), width = 200))
```

### 3.4.3 Naive peak calling
```{r rle_slice, eval=FALSE}
islands <- slice(cov, lower = 15)
islands[[1]]
```

### 3.4.4 Plot coverage for defined region

This chunk of code takes a lot of time, and often fails for the whole range. I did not got time to troubleshoot it. 
```{r plot_coverage, eval=FALSE}
library(ggbio)
chrLen <- c(length(cov$Chr1), length(cov$Chr2),length(cov$Chr3), length(cov$Chr4), length(cov$Chr5))
chr <- c("Chr1", "Chr2", "Chr3", "Chr4", "Chr5")
i= 'Chr1'
i = "Chr1"
for(i in chr){
   z <- as.character(i)
   z <- paste(z, ".png", sep="")
   png(z)
   q = 1
   myloc <- c(i, 1, chrLen[q])
   a <- readGAlignments(outpaths(args)[1], use.names=TRUE, param=ScanBamParam(which=GRanges(myloc[1], IRanges(as.numeric(myloc[2]), as.numeric(myloc[3])))))
   png(file='Chr_cov.png')
   autoplot(a, aes(color = strand, fill = strand), facets = strand ~ seqnames, stat = "coverage")
   dev.off()
   q <- q+1
 }

```

Instead, I have used following code to get coverage from rle object generated previously.
```{r plot_coverage_rle, eval=FALSE}
z <- GRanges("Chr1",IRanges(1,chrLen[1]))
xnum <- as.numeric(cov$Chr1[ranges(z)])
png(file='Chr1.png', res=300)
plot(xnum, type="l", col="blue", lwd=2, xlab='Index (bp)', ylab='Coverage')
dev.off()
```

![](Chr1.png)
<div> Coverage summary of chromosome 1</div></br>


## 3.5 Peak calling with MACS2

### 3.5.1 Merge BAM files of replicates prior to peak calling


Merging BAM files of technical and/or biological replicates can improve
the sensitivity of the peak calling by increasing the depth of read
coverage.

```{r merge_bams, eval=FALSE}
args <- systemArgs(sysma=NULL, mytargets="targets_bam.txt")
args_merge <- mergeBamByFactor(args, overwrite=TRUE)
writeTargetsout(x=args_merge, file="targets_mergeBamByFactor.txt", overwrite=TRUE)
```

### 3.5.2 Peak calling without input/reference sample

MACS2 can perform peak calling on ChIP-Seq data with and without input
samples [@Zhang2008-pc]. Here, I have used peak calling without input/reference sample.

```{r call_peaks_macs_noref, eval=FALSE}
args <- systemArgs(sysma="param/macs2_noinput.param", mytargets="targets_mergeBamByFactor.txt")
args <- systemArgs(sysma="param/macs2_noinput.param", mytargets="targets_mergeBamByFactor.txt")
sysargs(args)[1] # Command-line parameters for first FASTQ file
moduleload(module="python", envir=c("PATH", "LD_LIBRARY_PATH", "PYTHONPATH")) # Temp solution due to Python path change
moduleload(module="python")
runCommandline(args)

file.exists(outpaths(args))
writeTargetsout(x=args, file="targets_macs.txt", overwrite=TRUE)
```


The peak calling results from MACS2 are written for each sample to
separate files in the `results` directory. They are named after the corresponding
files with extensions used by MACS2.


### 3.5.3 Identify consensus peaks

The following example shows how one can identify consensus preaks among two peak sets sharing either a minimum absolute overlap and/or
minimum relative overlap using the `subsetByOverlaps` or `olRanges` functions, respectively. Note, the latter is
a custom function imported below by sourcing it.

```{r consensus_peaks, eval=FALSE}
source("http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/My_R_Scripts/rangeoverlapper.R")
peak_M1A <- outpaths(args)["AP1_1"]
peak_M1A <- as(read.delim(peak_M1A, comment="#")[,1:3], "GRanges")
peak_A1A <- outpaths(args)["C_1A"]
peak_A1A <- as(read.delim(peak_A1A, comment="#")[,1:3], "GRanges")
(myol1 <- subsetByOverlaps(peak_M1A, peak_A1A, minoverlap=1)) # Returns any overlap
myol2 <- olRanges(query=peak_M1A, subject=peak_A1A, output="gr") # Returns any overlap with OL length information
myol2[values(myol2)["OLpercQ"][,1]>=50] # Returns only query peaks with a minimum overlap of 50%
```

## 3.6 Annotate peaks with genomic context

### 3.6.1 Annotation with `ChIPpeakAnno` package

The following annotates the identified peaks with genomic context
information using the `ChIPpeakAnno` and `ChIPseeker` packages, respectively
[@Zhu2010-zo; @Yu2015-xu].

```{r chip_peak_anno, eval=FALSE}
library(ChIPpeakAnno); library(GenomicFeatures)
args <- systemArgs(sysma="param/annotate_peaks.param", mytargets="targets_macs_input.txt")
txdb <- loadDb("./data/TAIR10.sqlite")
#txdb <- makeTxDbFromGFF(file="data/tair10.gff", format="gff", dataSource="TAIR", organism="Arabidopsis thaliana")
#saveDb(txdb, file="./data/TAIR10.sqlite")
ge <- genes(txdb, columns=c("tx_name", "gene_id", "tx_type")) 
for(i in seq(along=args)) {
    peaksGR <- as(read.delim(infile1(args)[i], comment="#"), "GRanges")
    annotatedPeak <- annotatePeakInBatch(peaksGR, AnnotationData=genes(txdb))
    df <- data.frame(as.data.frame(annotatedPeak), as.data.frame(values(ge[values(annotatedPeak)$feature,])))
    write.table(df, outpaths(args[i]), quote=FALSE, row.names=FALSE, sep='\t')
}
writeTargetsout(x=args, file="targets_peakanno.txt", overwrite=TRUE)
```

The peak annotation results are written for each peak set to separate
files in the `results` directory. They are named after the corresponding peak
files with extensions specified in the `annotate_peaks.param` file, 
here `*.peaks.annotated.xls`.

### 3.6.2 Annotation with `ChIPseeker` package

Same as in previous step but using the `ChIPseeker` package for annotating the peaks.

```{r chip_peak_seeker, eval=FALSE}
library(ChIPseeker)
for(i in seq(along=args)) {
    peakAnno <- annotatePeak(infile1(args)[i], TxDb=txdb, verbose=FALSE)
    df <- as.data.frame(peakAnno)
    write.table(df, outpaths(args[i]), quote=FALSE, row.names=FALSE, sep="\t")
}
writeTargetsout(x=args, file="targets_peakanno.txt", overwrite=TRUE)
```

Summary plots provided by the `ChIPseeker` package. Here applied only to one sample
for demonstration purposes.

```{r chip_peak_seeker_plots, eval=FALSE}
peak <- readPeakFile(infile1(args)[1])
png(file='covplot.png')
covplot(peak, weightCol="X.log10.pvalue.")
dev.off()

png(file='peakHeatMap.png')
peakHeatmap(outpaths(args)[1], TxDb=txdb, upstream=1000, downstream=1000, color="red")
dev.off()

png(file='AvgProf.png')
plotAvgProf2(outpaths(args)[1], TxDb=txdb, upstream=1000, downstream=1000, xlab="Genomic Region (5'->3')", ylab = "Read Count Frequency")
dev.off()
```

![](covplot.png)
<div>ChIP peak distribution over all chromosomes</div></br>

![](peakHeatMap.png)
<div>Profile of peaks that align to flank sequences of TSS</div></br>


![](AvgProf.png)
<div>Peak distribution around TSS</div></br>

## 3.7 Count reads overlapping peaks

Counted number of reads on overlapping peaks. This will be useful in differential binding analysis.

```{r count_peak_ranges, eval=FALSE}
library(GenomicRanges)
args <- systemArgs(sysma="param/count_rangesets.param", mytargets="targets_macs_input.txt")
args_bam <- systemArgs(sysma=NULL, mytargets="targets_bam.txt")
bfl <- BamFileList(outpaths(args_bam), yieldSize=50000, index=character())
countDFnames <- countRangeset(bfl, args, mode="Union", ignore.strand=TRUE)
writeTargetsout(x=args, file="targets_countDF.txt", overwrite=TRUE)
```

## 3.8 Differential binding analysis

Differential binding analysis. It is possible to sort out sequences with low FDR/p-value.

```{r diff_bind_analysis, eval=FALSE}
args_diff <- systemArgs(sysma="param/rundiff.param", mytargets="targets_countDF.txt")
cmp <- readComp(file=args_bam, format="matrix") 
dbrlist <- runDiff(args=args_diff, diffFct=run_edgeR, targets=targetsin(args_bam), 
                    cmp=cmp[[1]], independent=TRUE, dbrfilter=c(Fold=0.75, FDR=.1))
writeTargetsout(x=args_diff, file="targets_rundiff.txt", overwrite=TRUE)
```


## 3.9 GO term enrichment analysis

The following performs GO term enrichment analysis for each annotated peak set.

### 3.9.1 Obtain gene-to-GO mappings

Before GO-mapping, it's needed to grab GO terms for our organism and store in a local database.
```{r go_mappings, eval=FALSE}
library("biomaRt")
listMarts() # To choose BioMart database
listMarts(host="plants.ensembl.org")
m <- useMart("plants_mart", host="plants.ensembl.org")
listDatasets(m)
m <- useMart("plants_mart", dataset="athaliana_eg_gene", host="plants.ensembl.org")
listAttributes(m) # Choose data types you want to download
go <- getBM(attributes=c("go_id", "tair_locus", "namespace_1003"), mart=m)
go <- go[go[,3]!="",]; go[,3] <- as.character(go[,3])
go[go[,3]=="molecular_function", 3] <- "F"; go[go[,3]=="biological_process", 3] <- "P"; go[go[,3]=="cellular_component", 3] <- "C"
go[1:4,]
dir.create("./data/GO")
write.table(go, "data/GO/GOannotationsBiomart_mod.txt", quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
catdb <- makeCATdb(myfile="data/GO/GOannotationsBiomart_mod.txt", lib=NULL, org="", colno=c(1,2,3), idconv=NULL)
save(catdb, file="data/GO/catdb.RData")
```

### 3.9.2 Run GO term enrichment analysis

GO-Enrichment
```{r go_enrich, eval=FALSE}
args <- systemArgs(sysma="param/macs2.param", mytargets="targets_bam_ref.txt")
args_anno <- systemArgs(sysma="param/annotate_peaks.param", mytargets="targets_macs_input.txt")
annofiles <- outpaths(args_anno)
gene_ids <- sapply(names(annofiles), function(x) unique(as.character(read.delim(annofiles[x])[,"geneId"])), simplify=FALSE)
load("data/GO/catdb.RData")
BatchResult <- GOCluster_Report(catdb=catdb, setlist=gene_ids, method="all", id_type="gene", CLSZ=2, cutoff=0.9, gocats=c("MF", "BP", "CC"), recordSpecGO=NULL)
```


## 3.10 Motif analysis

### 3.10.1 Parse DNA sequences of peak regions from genome

Enrichment analysis of known DNA binding motifs or _de novo_ discovery
of novel motifs requires the DNA sequences of the identified peak
regions. 

```{r parse_peak_sequences, eval=FALSE}
library(Biostrings); library(seqLogo); library(BCRANK)
args <- systemArgs(sysma="param/annotate_peaks.param", mytargets="targets_macs_input.txt")
rangefiles <- infile1(args)
for(i in seq(along=rangefiles)) {
    df <- read.delim(rangefiles[i], comment="#")
    peaks <- as(df, "GRanges")
    names(peaks) <- paste0(as.character(seqnames(peaks)), "_", start(peaks), "-", end(peaks))
    peaks <- peaks[order(values(peaks)$X.log10.pvalue., decreasing=TRUE)]
    pseq <- getSeq(FaFile("./data/tair10.fasta"), peaks)
    names(pseq) <- names(peaks)
    writeXStringSet(pseq, paste0(rangefiles[i], ".fasta")) 
}
```

## 3.11 Challenge Tasks

This part of the report provides my solution to challenge task.

### 3.11.1 Prioritize/rank peaks by FDR from differential binding analysis

```{r challenge_1, eval=FALSE}
diffBind <- read.table("./results/SRR038848_1.fastq_trim.gz_C.bam_macs2_peaks.edgeR.xls") #read edgeR output
diffBind.sorted <- diffBind[order(diffBind$AP1.C_FDR, diffBind$AP1.C_PValue),]
```


### 3.11.2 Parse peak sequences from genome sorted by lowest FDR/p-value

This codeblock ranks sequences from genome by lowest FDR/p-value
```{r parse_sorted_peaks, eval=FALSE}
peakFile <- 'results/SRR038848_1.fastq_trim.gz_C.bam_macs2_peaks.xls.fasta'
sortedPeaks <-readDNAStringSet(peakFile)[rownames(diffBind.sorted)]
writeXStringSet(sortedPeaks, 'results/sortedPeaks.fasta')
```


### 3.11.3 Enrichment analysis

I use this code for motif enrichment. I needed to compute background frequency of promoter region sequences (1000bp upstream)

```{r motif_enrichment, eval=FALSE}
library(MotifDb)
library(Biostrings)
library(PWMEnrich)
library(GenomicFeatures)
library(ShortRead)
library(BSgenome)

peak_seq <- readDNAStringSet('results/sortedPeaks.fasta',format='fasta')
athaliana_motifs <- as.list(query(MotifDb, 'athaliana'))

# PWMEnrich needs pfm as integer matrix
pfm_int_mat <- function(pfm){
        pfm <- as.integer(as.matrix(pfm)*10000000)
        pfm <- matrix(pfm, nrow=4)
        rownames(pfm) <- c('A', 'C', 'G', 'T')
        colnames(pfm) <- 1:dim(pfm)[2]
        return(pfm)
}

# Convert MotifDb athaliana motifs into integer matrix
pfm <- sapply(athaliana_motifs, pfm_int_mat)

# Need promoter sequence for calculating background frequency
txdb <- loadDb("./data/TAIR10.sqlite")
promoters <- promoters(genes(txdb), upstream = 1000, downstream = 0)
#promoters <- sample(promoters, 5000, replace=FALSE)
genomeFa <- readDNAStringSet('data/tair10.fasta')
#promoter_seq <- getSeq(genomeFa, promoters)
promoter_seq<- getSeq(genomeFa, promoters[seqnames(promoters)=='Chr1' |
                      seqnames(promoters)=='Chr2' | seqnames(promoters)=='Chr3' |
                      seqnames(promoters)=='Chr4' | seqnames(promoters)=='Chr5'])

# Motif enrichment needs LogN background
#sample_promoter_seqs <- sample(promoter_seq, 1000, replace=FALSE)
#PWMLogn.tair10.MotifDb.Athal<- makePWMLognBackground(clean(promoter_seq), pfm)
#save(PWM_bg, file='PWMLogn.tair10.MotifDb.Athal')
load('Athal.TAIR10.PWM.bg')

# Motif enrichment in multiple sequences
res <- motifEnrichment(peak_seq[1:100], PWM_bg)
#report <- groupReport(res)
#save(report, file='100_peak.report')
#load('all_peak.report')
load('all_peak.report')
report

png(filename='100group_enriched_motif.png')

plot(report[1:10], fontsize=7, id.fontsize=5)
dev.off()
```

Top 10 motif enrichment hits based on all peak sequences:

```{r top_10_on_all}
load('all_peak.report')
report
```

![](group_enriched_motif.png)
<div>Top ten enriched-motifs in the whole peak-set</div></br>

We can see TCP is here, which is a basic helix-loop-helix transcription factor. It also encodes a transcription factor AtTCP14 that regulates seed germination.

Kaufmann et al., 2010 also find TCPs overrepresented target genes of the MADS domain transcription factor AP1


Top 10 motif enrichment hits based on 100 top peak sequences (low FDR/p-value):

```{r top_10_on_100}
load('100_peak.report')
report
```

![](100group_enriched_motif.png)
<div>Top ten enriched motiffs in the top-100 peak-set</div></br>
Now, we can see in this list, it's highly enriched by AP-family transcription factors. It also corroborate with the paper that AP1-binding have some self-feedback loop mechanism.

### 3.11.4 Comparing differentially expressed genes from the paper and differentially binding genes from our analysis
```{r DE_DB_comparison, eval=FALSE}
ann_table <- read.table("./results/SRR038848_1.fastq_trim.gz_C.bam_macs2_peaks.annotated.xls", sep="\t", header = T)
venn_targ <- unique(ann_table$geneId[ann_table$fold_enrichment >= 1.8]) # this cutoff used by the paper
xl <- read.table("/bigdata/gen242/tkata002/5_30/gen242_2018_ChIP-Seq1/chipseq/results/genes.csv") # from theo

l1 <-levels(xl$V1)
l2 <-levels(ann_table$geneId)

intersect(l1,l2) # Identify which genes overlap
sum(l1%in%l2) # Count of overlapping genes

setlist5 <- list(Kaufmann_DE_list = l1, This_study= l2)
OLlist5 <- overLapper(setlist=setlist5, sep="_", type="vennsets")
png("venn_plot.png")
vennPlot(OLlist5, mymain="", mysub="", colmode=2, ccol=c("blue", "red"))
dev.off()
```

![](venn_plot.png)
<div>In present setting, Theo got five genes that are common in DE-genes from the paper and differentially-biding analysis of this project</div></br>

The very small match with these two dataset is not unsurprising, because data analysis parameters are very different between the paper and this work. Howevere, there is opportunity to optimize the parametric settings.


### 3.11.5 TSS distance distribution
```{r TSS_hist, eval=FALSE}
png(file='TSShist.png')
hist(ann_table$distanceToTSS, xlab = "Distance to transcription start site", main = "All Peaks", freq = T )
dev.off()
```

![](TSShist.png)
<div>Distance to TSS </div></br>

This indicates preference of the motifs to binding just before the transcription start site (TSS).

# 4. Version Information

```{r sessionInfo}
sessionInfo()
```


# 5. Funding

This project was supported by funds from the National Institutes of
Health (NIH) and the National Science Foundation (NSF).

# 6. References

