---
title: "Kataras_ChIP_Seq1"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Abstract
# The APETALA1(AP1) gene of Arabadopsis thaliana is relevant for floral formation pathways and for repressor relationships in a broader context. Specifically, the AP1 gene supresses floral supressors, promoting flower formation.

##Introduction
# At the time this paper was published, little was known about the molecular mechanisms of the relulatory proteins involved in floral initiation. The model organism Arabadopsis Thaliana was used in this study to elucidate the interactions of the AP1 gene in the flowering pathways.

# This study determined the place of AP1 in this pathway partly through combination with its paralog, CAULIFLOWER. Noting that when both genes were mutated, no floral fomation was observed and in place of flowers an overgrowth of inflorescense meristems.

# Transciption factors which were previously identified which bind to AP1's promoter region, but the proteins downstream, those effected by AP1 were not yet identified. So, the researchers used a combination of microarray and ChIP seq data to find the targets of the AP1 gene at multiple time points after AP1 introduction on a genome-wide scale.

##Methods
# A brief overview of the methods used in ChIP Seq are as follows:
# Induce a condition to effect gene expression, here the introduction of AP1
# Shear the DNA and pulldown the AP1 bound fragments with AP1 specific antibodies
# Seperate the proteins from the DNA and sequence the DNA fragments, here via Illumina sequencing

#ChIP Seq workflow:

#The first step of the ChIP Seq workflow is to call the neccessary libraries and initialize the work environment


```{r libraries}
suppressPackageStartupMessages({
    library(systemPipeR)
    library(BiocParallel)
    library(Biostrings)
    library(Rsamtools)
    library(GenomicRanges)
    library(ggplot2)
    library(GenomicAlignments)
    library(ShortRead)
    library(ape)
    library(systemPipeRdata)
    library(systemPipeR)
    
})
```


# The workspace in intialized from the shared github directory to ensure all files are up to date
```{r git_clone}
system("git clone https://github.com/girke-class/gen242_2018_ChIP-Seq1")
```

# Specifying and displaying some of the targets file holding the ChIP seq reads
```{r targets}
targetspath <- system.file("./targets_chip.txt", package="systemPipeR")
targets <- read.delim(targetspath, comment.char = "#")
targets[1:4,-c(5,6)]
```

# Before reading in sequence read files, it can be advisable to quality filter the reads to exclude low quality reads that decrease the certainty of subsequent analysis. However, for our workflow, we found that filtering the reads excluded too many and negatively impacted our analysis, so the following step is commented out.
```{r preprocess}
## ----proprocess_reads, eval=FALSE, messages=FALSE, warning=FALSE, cache=TRUE----
# args <- systemArgs(sysma="param/trim.param", mytargets="targets_chip.txt")
# system("cat ")
# 
# #### quality control Make more permissive! 
# filterFct <- function(fq, cutoff=20, Nexceptions=0) {
#   qcount <- rowSums(as(quality(fq), "matrix") <= cutoff)
#   fq[qcount <= Nexceptions] # Retains reads where Phred scores are >= cutoff with N exceptions
#   }
#   preprocessReads(args=args, Fct="filterFct(fq, cutoff=20, Nexceptions=0)", batchsize=100000)
#   writeTargetsout(x=args, file="targets_chip_trim.txt", overwrite=TRUE)

```

# The next step in the ChIP Seq workflow is to align the collected reads. We chose to use Bowtie 2 for this step.
```{r bw2_align}
args <- systemArgs(sysma="param/bowtieSE.param", mytargets="targets_chip.txt")
runCommandline(args)
writeTargetsout(x=args, file="targets_bam.txt", overwrite=TRUE)#**need to move this down below runCommandline???

```

# We check that the .bam files from the alignment were successfully created
```{r check_bam}
system("git clone https://github.com/girke-class/gen242_2018_ChIP-Seq1")
```


# Write out an Excel file with alignent stats. compare to  alignStatsTheo.xls which was the result of alignment on the filtered data
```{r align_stats}
read_statsDF <- alignStats(args=args)
write.table(read_statsDF, "./results/alignStats.xls", row.names=FALSE, quote=FALSE, sep="\t")
read.delim("results/alignStats.xls")

read.delim("results/alignStatsTheo.xls")

```

# Next we create a link to the aligned bam files for graphing
```{r bam_link}
symLink2bam(sysargs=args, htmldir=c("~/.html/", "gen242/"),
            urlbase="http://biocluster.ucr.edu/~tkata002/", urlfile="results/IGVurl.txt")

```

# Creating objects to store coverage information. This will show us where the reads from our **only using one of the 7 outpaths args??
```{r cov}
library(rtracklayer); library(GenomicRanges); library(Rsamtools); library(GenomicAlignments)
aligns <- readGAlignments(outpaths(args)[1])
cov <- coverage(aligns)
```

# Combines the bam files from our multiple fastq files from the multiple biological and technical replicates. All files but AP1_1 had a technical replicate. There was 1 biological replicate for AP1 and the control. 
```{r git_clone}
system("git clone https://github.com/girke-class/gen242_2018_ChIP-Seq1")
```



# Next we use the aligned coverage information to call peaks, identifying parts of the genome which were frequently bound by AP1 protein. Peak calling is done without reference, this uses the control sample as reference to normalize by the random background.
# Had to do this in R on cluster directly line by line.**
```{r call_peaks}
args <- systemArgs(sysma="param/macs2_noinput.param", mytargets="targets_mergeBamByFactor.txt")
moduleload(module="python", envir=c("PATH", "LD_LIBRARY_PATH", "PYTHONPATH")) # Temp solution due to Python path change
moduleload(module="python")
runCommandline(args)
#file.exists(outpaths(args))
writeTargetsout(x=args, file="targets_macs.txt", overwrite=TRUE)

```


# Now we define consensus peaks from the peak calling
# This calls peaks.annotated.xls but we have not annotated yet????*** get from arafat
```{r consensus_peaks}
source("http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/My_R_Scripts/rangeoverlapper.R")
peak_C <- outpaths(args)["C_1A"] ####currently only has "C_1A" now!!!!
peak_C <- as(read.delim(peak_C, comment="#")[,1:3], "GRanges") ##!!! why 1:3????

```


# We take the consensus peaks and use the CHiPpeakAnno package to determine relavant genome infomation about the peaks, connecting our peaks from the workflow to the larger biological context.  
```{r peak_anno}
library(ChIPpeakAnno); library(GenomicFeatures)
args <- systemArgs(sysma="param/annotate_peaks.param", mytargets="targets_macs.txt")
txdb <- loadDb("./data/tair10.sqlite")
#txdb <- makeTxDbFromGFF(file="data/tair10.gff", format="gff", dataSource="TAIR", organism="Arabidopsis thaliana")
ge <- genes(txdb, columns=c("tx_name", "gene_id", "tx_type"))
for(i in seq(along=args)) {
    peaksGR <- as(read.delim(infile1(args)[i], comment="#"), "GRanges")
    annotatedPeak <- annotatePeakInBatch(peaksGR, AnnotationData=genes(txdb))
    df <- data.frame(as.data.frame(annotatedPeak), as.data.frame(values(ge[values(annotatedPeak)$feature,])))
    write.table(df, outpaths(args[i]), quote=FALSE, row.names=FALSE, sep="\t")
}
writeTargetsout(x=args, file="targets_peakanno.txt", overwrite=TRUE)
```


# Or we can use the ChIPseeker package to identify the nearest genes to our peaks, identifying the most likely targets of the AP1 binding
```{r peak_seeker}
library(ChIPseeker)
for(i in seq(along=args)) {
    peakAnno <- annotatePeak(infile1(args)[i], TxDb=txdb, verbose=FALSE)
    df <- as.data.frame(peakAnno)
    write.table(df, outpaths(args[i]), quote=FALSE, row.names=FALSE, sep="\t")
}
writeTargetsout(x=args, file="targets_peakanno.txt", overwrite=TRUE)
```






# This counts the peaks ranges, allowing us to see where peaks are most prevalent by reading counting iteratively over "several range sets"**. **overlapping??
```{r peak_range_count}
library(GenomicRanges)
args <- systemArgs(sysma="param/count_rangesets.param", mytargets="targets_macs.txt")
args_bam <- systemArgs(sysma=NULL, mytargets="targets_bam.txt")
bfl <- BamFileList(outpaths(args_bam), yieldSize=50000, index=character())
countDFnames <- countRangeset(bfl, args, mode="Union", ignore.strand=TRUE)
writeTargetsout(x=args, file="targets_countDF.txt", overwrite=TRUE) #XX
```

# Differential binding analysis. Creates edgeR.xls files which contain differential expression information. The EdgeR package determines differential expression analysis from our ChIP data peak sequence enrichment.*
```{r edgeR}
args_diff <- systemArgs(sysma="param/rundiff.param", mytargets="targets_countDF.txt")
cmp <- readComp(file=args_bam, format="matrix")
dbrlist <- runDiff(args=args_diff, diffFct=run_edgeR, targets=targetsin(args_bam),
                    cmp=cmp[[1]], independent=TRUE, dbrfilter=c(Fold=.2, FDR=.05))
writeTargetsout(x=args_diff, file="targets_rundiff.txt", overwrite=TRUE)
```

# Now that we have our differentially expressed genes, we use them for Gene Ontology analysis, to determine if there are any trends in the type of genes effected by AP1 binding ***** mine didnt work!!
```{r go_enrich}
args <- systemArgs(sysma="param/macs2.param", mytargets="targets_bam_ref.txt")
args_anno <- systemArgs(sysma="param/annotate_peaks.param", mytargets="targets_macs.txt")
annofiles <- outpaths(args_anno)

load("data/GO/catdb.RData")
BatchResult <- GOCluster_Report(catdb=catdb, setlist=gene_ids, method="all", id_type="gene", CLSZ=2, cutoff=0.9,  recordSpecGO=NULL) #gocats=c("MF", "BP", "CC"),
head(BatchResult)
```

# Next we parse the peak sequences we foudn earlier from the genome, this completes the second goal of our challenge project ** may need to get this from arafat
```{r parse_peaks}
library(Biostrings); library(seqLogo); library(BCRANK)
args <- systemArgs(sysma="param/annotate_peaks.param", mytargets="targets_macs.txt")
rangefiles <- infile1(args)
for(i in seq(along=rangefiles)) {
  
  i = 1
    df <- read.delim(rangefiles[i], comment="#")
    peaks <- as(df, "GRanges")
    names(peaks) <- paste0(as.character(seqnames(peaks)), "_", start(peaks), "-", end(peaks))
    #peaks <- peaks[order(values(peaks)$X.log10.pvalue.,values(peaks)$, decreasing=TRUE)] this is different and I dont know why
    peaks <- peaks[order(values(peaks)$X.log10.pvalue., decreasing=TRUE)]
    pseq <- getSeq(FaFile("./data/tair10.fasta"), peaks)
    names(pseq) <- names(peaks)
    writeXStringSet(pseq, paste0(rangefiles[i], "Theo.fasta"))
}
head(peaks)
```

# This next section completes the first challenge project objective:"Prioritize/rank peaks by FDR from differential binding analysis"
# We use the order commands within R, and the EdgeR output file to sort the peaks by FDR and then p value. Sorting by smallest FDR and then by smallest p value for each FDR threshold. ** arafat**
```{r git_clone}
##############Challenge project part 1, FDR "Prioritize/rank peaks by FDR from differential binding analysis"
edgeR_table <- read.table("./results/SRR038848_1.fastq_trim.gz_C.bam_macs2_peaks.edgeR.xls")
####sorting by smallest FDR and then smallest p value
edgeR_table <- read.table("./results/SRR038848_1.fastq_trim.gz_C.bam_macs2_peaks.edgeR.xls")
####sorting by smallest FDR and then smallest p value
edgeR_table_sorted <- edgeR_table[order(edgeR_table$AP1.C_FDR, edgeR_table$AP1.C_PValue),]
head(edgeR_table_sorted)

```


# This next code completes the second objective of our challenge project, parsing the peak sequences from the genome.
# Presenting the peak sequences from  Macs2 peak calling. These are the sequences which were highly enriched within the set of sequences returned from the Illumina sequencing
# The sequences are saved as character strings with names of each being the chromosome and range designation.
```{r parse_peaks2}
chr_test <- readDNAStringSet("./results/SRR038848_1.fastq_trim.gz_C.bam_macs2_peaks.xls.fasta")
peak_seqs <- as.character(chr_test)
head(peak_seqs)

```

# The third challenge project objective is to determine which motifs show the highest enrichment given a variety of Position Weight Matrix sets from the JASPAR database. We used the new and robust PWMEnrich package to accomplish this. The difficult part was creating the random background for arabadopsis, as the package did not have a preset. To do this we used ****, and ultimatly *** was the PWM with greatest enrichment for our motifs and *** was the most enriched motif from the peak sequences  ** arafat
```{r motif_enrich}





```







# The workspace in intialized from the shared github directory to ensure all files are up to date
```{r git_clone}
system("git clone https://github.com/girke-class/gen242_2018_ChIP-Seq1")
```




## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.




#Results

# Plots coverage for all chromosomes using the merged bam. doesnt work currently**
```{r cov_plots}

args <- systemArgs(sysma="param/bowtieSE.param", mytargets="targets_chip_trim.txt")

library(ggbio)
 chr <- c("Chr1", "Chr2", "Chr3", "Chr4", "Chr5")
 chr_length <- c(30000000,21000000,25000000,20000000,30000000)# chromosome lengths for TAIR
 
 i = "Chr1"
 for(i in chr){
   z <- as.character(i)
   z <- paste(z, ".pdf", sep="")
   pdf(z)
   q = 1
   myloc <- c(i, 1, chr_length[q])
   a <- readGAlignments(outpaths(args)[1], use.names=TRUE, param=ScanBamParam(which=GRanges(myloc[1], IRanges(as.numeric(myloc[2]), as.numeric(myloc[3])))))
   autoplot(a, aes(color = strand, fill = strand), facets = strand ~ seqnames, stat = "coverage")
   dev.off()
   q <- q+1
 }
```



# We can use the peak seeker program to create graphs of the peaks to see where our coverage piled up most significantly . will this show graph?**
```{r peak_seeker_plots}

pdf("peakHeatmap.pdf")
peak <- readPeakFile(infile1(args)[1])
covplot(peak, weightCol="X.log10.pvalue.")
peakHeatmap(outpaths(args)[1], TxDb=txdb, upstream=1000, downstream=1000, color="red")

dev.off()

plotAvgProf2(outpaths(args)[1], TxDb=txdb, upstream=1000, downstream=1000, xlab="Genomic Region (5'->3')", ylab = "Read Count Frequency")
```




# Comparing differentially expressed gene lists for the paper and our workflow outcomes. We were interested in determining how many of our putative target genes for differential AP1 binding matched those identified in the paper
```{r venn_targets_workflow}
ann_table <- read.table("./results/SRR038848_1.fastq_trim.gz_C.bam_macs2_peaks.annotated.xls", sep="\t", header = T)

#ann_table2 <- ann_table[ann_table$geneChr <= 5,]# filters out chrM and chrC, as the chromosomes are listed 1 to 7
#dim(ann_table)
#dim(ann_table2)#the difference between including and excluding the last 2 chromosomes is only 20 genes at this point of over 1200
venn_targ <- (unique(genes[ann_table$fold_enrichment >= 1.8]))
#length(venn_targ)
```

#the paper values are the table 3 data from paper, the differentially expressed genes "249 genes (which we refer to as high-confidence targets) showed robust (>1.8-fold) differential expression" from microarray data.
#our genes by comparison are found to be differentially expressed by EdgeR, using the same 1.8 fold cutoff. 
```{r venn_targets_paper}
xl <- read.table("/bigdata/gen242/tkata002/5_30/gen242_2018_ChIP-Seq1/chipseq/results/genes.csv")
```


#finding out how many genes overlap between paper and our own workflow, idenify gene ids, count and create venn diagrams
```{r venn_overlaps}
l1 <-levels(xl$V1)
l2 <-levels(ann_table$geneId)

intersect(l1,l2) # Identify which genes overlap
sum(l1%in%l2) # Count of overlapping genes

library(systemPipeR)
setlist5 <- list(Paper_diff_exp = l1, Our_diff_enrich= l2)

OLlist5 <- overLapper(setlist=setlist5, sep="_", type="vennsets")
pdf("venn_plot.pdf")
vennPlot(OLlist5, mymain="", mysub="", colmode=2, ccol=c("blue", "red"))
dev.off()
```
#though the number of genes differntially expressed and enriched was strikingly similar, the overlap between the two sets was rather small, sharing just 5 genes, or 2% of the sets. This is somewhat troubling, but could be a result of differnt peak caller programs being utilized in our workflow and in the paper.



# An interesting result of ours that mirrors the paper is the transcipt type breakdown. We found that more than 50% of the total peaks were nearest a gene that coded for an mRNA, indicating that AP1 effects largely protein coding genes, as would be expected of a transcirption factor upregualting floral formation.
#### transcript type breakdown
```{r tx_type}
names(ann_table)
tx_breakdown <- barplot(prop.table(table(ann_table$tx_type)))
plot(tx_breakdown)  
```

# Finally, the most important figure in my opinion of this paper was 1F, adressing where the AP1 protein bound most commonly and it showed distinct preference for binding just upstream of the transcription start site. 
# our data from the workflow did not show as striking of a distinction as the graph in the paper, but there was a slight preference for binding just upstream of the TSS.
```{r TSS_hist}
hist(ann_table$distanceToTSS, xlab = "Distance to transcription start site", main = "All Peaks", freq = T )


```


#Conclusions

#As the paper found, the binding preference of AP1 seems to be just upstream of the TSS, a behavior that reveals its role as a transcitpion factor. Our ChIP Seq workflow did not include the comparison to SEP1 ** which AP1 inhibits, to induce flower formation, but the interaction between the two revealed the molecular mechanism of inhibiting the inhibitors through which AP1 causes floral formation.


#Acknowledgments

#References

